### PR TITLE
fix failed back transfer

### PR DESCRIPTION
**Pull request by taitruong**
# Bug description

- an NFT gets transferred back from chain B to A
- in case of failure (e.g. caused by incoming proxy or callback) nothing should happen
- instead states for outgoing channel are not reverted.

Task to reproduce:

0. do NOT whitelist channel in incoming proxy on chain A -> this will cause an error on back transfer
1. transfer NFT from A to B
2. transfer NFT back from B to A

# 1. NFT (forward) transferred from chain A to B

Outcoume:

- chain A: NFT escrowed by ICS721
- chain B: NFT minted on newly, instantiated collection
- state changes:
  - entry added on chain A: `OUTGOING_CLASS_TOKEN_TO_CHANNEL` ("returning to source" identifier for back transfer)
  - entry added on chain B: `INCOMING_CLASS_TOKEN_TO_CHANNEL`

# 2. Transfer NFT back from B to A

Outcome:

- expected outcome on failure:
  - NFTs unchanged on chain A (escrowed) and B (holder)
  - no state changes
- actual outcome on failure:
  - NFTs unchanged (this is correct) on both chains
  - states not reverted

Reason here states are updated in contract, but all other actions are handled in sub messages. So if sub messages fails, states in ics721 are not reverted. Reason here is, that on ibc receive TX must not fail for relaying Ack success or error.

What happened next: I WLed incoming proxy and did another back transfer, this time successfully, but with an unexpected result:

- chain B: NFT burned
- chain A: 2 collections:
  - NFT minted on newly, instantiated collection
  - "source/OG" NFT still escrowed by ICS721

Reason here is because of previous failure, removal of entry in outgoing channel got not reverted. So on next transfer ics721 on chain A:

- does not recognise NFT as a back transfer
- NFT on correct collection stays escrowed by ics721
- a new collection is instanted
- NFT is mint on new collection

This PR fixes this, by moving state changes for incoming and outgoing channel, away from ics721 contract to a dedicated sub message. So if other sub messages like callback or incoming proxy fails, this submessage is not executed.

### CODE REVIEW

File: packages/ics721/src/ibc_packet_receive.rs [hunk 1]
```diff
@@ -94,57 +121,117 @@ pub(crate) fn receive_ibc_packet(
                         // We previously sent this NFT out on this
                         // channel. Unlock the local version for the
                         // receiver.
-                        OUTGOING_CLASS_TOKEN_TO_CHANNEL.remove(deps.storage, key);
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:97]: This here is the bug: state change is done here in ics721

File: packages/ics721/src/ibc_packet_receive.rs [hunk 2]
```diff
@@ -94,57 +121,117 @@ pub(crate) fn receive_ibc_packet(
...
+                        redemption_or_create.0.push(token_id);
+                        return Ok(redemption_or_create);
                     }
                 }
                 // It's not something we've sent out before => make a
                 // new NFT.
-                let local_prefix = get_endpoint_prefix(&packet.dest);
-                let local_class_id = ClassId::new(format!("{}{}", local_prefix, data.class_id));
-
-                INCOMING_CLASS_TOKEN_TO_CHANNEL.save(
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:110]: This is the other bug: state change is done also in ics721.

File: packages/ics721/src/ibc_packet_receive.rs [hunk 3]
```diff
@@ -154,190 +241,49 @@ pub(crate) fn receive_ibc_packet(
...
-    };
-
-    let incoming_proxy_msg =
-        get_incoming_proxy_msg(deps.storage, packet.clone(), cloned_data.clone())?;
-    let submessage = action_aggregator.into_submessage(
-        env.contract.address,
-        receiver,
-        callback_msg,
-        incoming_proxy_msg,
-    )?;
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:189]: Here is all the main logic: submessage itself contains a list of messags for:
- action aggregrator: creates a callback msg with a list of more messages for create or redeemining NFTs
- optional incoming proxy handling
- optional callback handling

File: packages/ics721/src/ibc_packet_receive.rs [hunk 4]
```diff
@@ -62,25 +31,83 @@ pub(crate) fn receive_ibc_packet(
...
+) -> Result<(bool, (WasmMsg, WasmMsg)), ContractError> {
+    let token_count = data.token_ids.len();
+    let redemption_or_create = data
         .token_ids
         .into_iter()
         .zip_optional(data.token_uris)
         .zip_optional(data.token_data)
         .try_fold(
-            Vec::<Action>::with_capacity(token_count),
-            |mut messages, ((token_id, token_uri), token_data)| -> StdResult<_> {
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:80]: old logic is super complicated for create a  callaback submessage, using Action and Action Aggregator

File: packages/ics721/src/ibc_packet_receive.rs [hunk 5]
```diff
@@ -62,25 +31,83 @@ pub(crate) fn receive_ibc_packet(
...
+    // sub message holds 2 to 4 messages:
+    // - one message for voucher creation or redemption, another message for updating incoming or outgoing channel
+    let (is_redemption, voucher_and_channel_messages) = create_voucher_and_channel_messages(
+        deps.as_ref(),
+        env.clone(),
+        data.clone(),
+        maybe_local_class_id,
+        local_class_id.clone(),
+        packet.clone(),
+    )?;
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:53]: new logic leads to same result, by creating required messages - without using temporary action and aggregator objects

File: packages/ics721/src/msg.rs [hunk 6]
```diff
@@ -63,6 +63,10 @@ pub enum CallbackMsg {
         /// Information about the vouchers been redeemed.
         redeem: VoucherRedemption,
     },
+    /// Redeem all entries in outgoing channel.
+    RedeemOutgoingChannelEntries(Vec<(ClassId, TokenId)>),
+    /// Save all entries in incoming channel.
+    AddIncomingChannelEntries(Vec<((ClassId, TokenId), String)>),
```
Comment (taitruong) [inline packages/ics721/src/msg.rs:69]: NEW: handle state changes for incoming and outgoing channel in separate callback message

File: packages/ics721/src/ibc_packet_receive.rs [hunk 7]
```diff
@@ -62,25 +31,83 @@ pub(crate) fn receive_ibc_packet(
...
+fn create_voucher_and_channel_messages(
+    deps: Deps,
+    env: Env,
+    data: NonFungibleTokenPacketData,
+    maybe_local_class_id: Option<&str>,
+    local_class_id: ClassId,
+    packet: IbcPacket,
+) -> Result<(bool, (WasmMsg, WasmMsg)), ContractError> {
+    let token_count = data.token_ids.len();
+    let redemption_or_create = data
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:98]: New logic is less (but still) complex, creates directly and return messages.

File: packages/ics721/src/ibc_packet_receive.rs [hunk 8]
```diff
@@ -94,57 +121,117 @@ pub(crate) fn receive_ibc_packet(
...
+                .into_iter()
+                .map(|token_id| (local_class_id.clone(), token_id))
+                .collect();
+            let redeem_outgoing_channels_msg = WasmMsg::Execute {
+                contract_addr: env.contract.address.to_string(),
+                msg: to_json_binary(&ExecuteMsg::Callback(
+                    CallbackMsg::RedeemOutgoingChannelEntries(redeem_outgoing_channels),
+                ))?,
+                funds: vec![],
+            };
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:175]: And adds either CallbackMsg::RedeemOutgoingChannelEntries ...

File: packages/ics721/src/ibc_packet_receive.rs [hunk 9]
```diff
@@ -94,57 +121,117 @@ pub(crate) fn receive_ibc_packet(
...
+                    )
+                })
+                .collect();
+            let add_incoming_channels_msg = WasmMsg::Execute {
+                contract_addr: env.contract.address.to_string(),
+                msg: to_json_binary(&ExecuteMsg::Callback(
+                    CallbackMsg::AddIncomingChannelEntries(add_incoming_channels),
+                ))?,
+                funds: vec![],
+            };
```
Comment (taitruong) [inline packages/ics721/src/ibc_packet_receive.rs:207]: ... CallbackMsg::AddIncomingChannelEntries which then will be added to sub message.

File: ts-relayer-tests/src/ics721.spec.ts [hunk 10]
```diff
@@ -311,95 +391,366 @@ test.serial(
...
+- osmo before: ${JSON.stringify(beforeOsmoIncomingClassTokenToChannelList)}
+- osmo after: ${JSON.stringify(afterOsmoIncomingClassTokenToChannelList)}`
+  );
+  t.deepEqual(
+    beforeOsmoNftContractsToClassIdList,
+    afterOsmoNftContractsToClassIdList,
+    `nft contracts must be unchanged:
+- osmo before: ${JSON.stringify(beforeOsmoNftContractsToClassIdList)}
+- osmo after: ${JSON.stringify(afterOsmoNftContractsToClassIdList)}`
+  );
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:536]: This is the ts relayer test, checking states before and after a failed back transfer

File: ts-relayer-tests/src/ics721.spec.ts [hunk 11]
```diff
@@ -311,95 +391,366 @@ test.serial(
...
+
+  // assert NFT on chain A is returned to owner
+  tokenOwner = await ownerOf(wasmClient, wasmCw721, tokenId);
+  t.is(wasmAddr, tokenOwner.owner);
+
+  // ==== test transfer NFT to osmo chain via channel WLed ONLY on osmo incoming proxy and back to wasm chain ====
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:542]: This test reproduces back transfer on failure and success:

File: ts-relayer-tests/src/ics721.spec.ts [hunk 12]
```diff
@@ -311,95 +391,366 @@ test.serial(
...
+  t.is(wasmIcs721, tokenOwner.owner);
+  // assert NFT on chain B is owned by osmoAddr
+  osmoClassId = `${onlyOsmoIncomingChannel.channel.dest.portId}/${onlyOsmoIncomingChannel.channel.dest.channelId}/${t.context.wasmCw721}`;
+  osmoCw721 = await osmoClient.sign.queryContractSmart(osmoIcs721, {
+    nft_contract: { class_id: osmoClassId },
+  });
+  tokenOwner = await ownerOf(osmoClient, osmoCw721, tokenId);
+  t.is(osmoAddr, tokenOwner.owner);
+
+  // test back transfer NFT to wasm chain, where onlyOsmoIncomingChannel is not WLed on wasm chain
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:615]: test failed back transfer

File: ts-relayer-tests/src/ics721.spec.ts [hunk 13]
```diff
@@ -311,95 +391,366 @@ test.serial(
...
-    t.log("relaying packets");
+  // assert after failed relay, NFT on chain B is returned to owner
+  allNFTs = await allTokens(osmoClient, osmoCw721);
+  t.true(allNFTs.tokens.length === 1);
+  // assert NFT is returned to sender on osmo chain
+  tokenOwner = await ownerOf(osmoClient, osmoCw721, tokenId);
+  t.is(osmoAddr, tokenOwner.owner);
 
-    info = await otherChannel.link.relayAll();
+  // ==== WL channel on wasm chain and test back transfer again ====
```
Comment (taitruong) [inline ts-relayer-tests/src/ics721.spec.ts:656]: migrate and WL channel for testing succesful back transfer, now NFT is unescrowd on existing collection - and NOT minted on a new collection

**Review by shanev — APPROVED** [review]

**Review by humanalgorithm — APPROVED** [review]
MrT and I went over these changes on a call

File: packages/ics721/src/ibc_packet_receive.rs [hunk 14]
```diff
@@ -62,25 +31,83 @@ pub(crate) fn receive_ibc_packet(
...
+    let (is_redemption, voucher_and_channel_messages) = create_voucher_and_channel_messages(
+        deps.as_ref(),
+        env.clone(),
+        data.clone(),
+        maybe_local_class_id,
+        local_class_id.clone(),
+        packet.clone(),
+    )?;
+    // - one optional incoming proxy message
+    let incoming_proxy_msg =
```
Comment (humanalgorithm) [inline packages/ics721/src/ibc_packet_receive.rs:55]: thanks for refactoring this out to functions 👍

**Review by jhernandezb — APPROVED** [review]

**Pull request merged**